### PR TITLE
Send more detailed metrics from statsd reporter

### PR DESCRIPTION
**Pull request by romcheg**
This patch adds the same metrics to statsd reporter as
Graphite reporter sends.

### CODE REVIEW

File: metrology/reporter/statsd.py [hunk 1]
```diff
@@ -101,11 +128,9 @@ def _send_tcp(self):
     def _send_udp(self):
         if len(self.batch_buffer):
             if sys.version_info[0] > 2:
-                self.socket.sendto(bytes(self.batch_buffer, 'ascii'),
-                                   (self.host, self.port))
+                self.socket.sendto(self.batch_buffer, (self.host, self.port))
```
Comment (mkurek) [inline metrology/reporter/statsd.py]: there is no difference between py2 and py3 now - shouldn't you use `bytes` here?

File: metrology/reporter/statsd.py [hunk 2]
```diff
@@ -51,40 +55,63 @@ def socket(self):
...
 
         self._send()
 
-    def send_metric(self, name, mtype, metric, keys):
-        value = getattr(metric, keys, None)
-        if value:
-            self._buffered_send_metric(name, value, mtype)
-
-    def _buffered_send_metric(self, name, value, mtype):
+    def send_metric(self, name, mtype, metric, keys, snapshot_keys=[]):
```
Comment (mkurek) [inline metrology/reporter/statsd.py]: don't use list as default arg

File: metrology/reporter/statsd.py [hunk 3]
```diff
@@ -51,40 +55,63 @@ def socket(self):
...
+        base_name = re.sub(r"\s+", "_", name)
         if self.prefix:
-            name = '{0}.{1}'.format(self.prefix, name)
+            base_name = "{0}.{1}".format(self.prefix, base_name)
 
-        self.batch_count += 1
-        if mtype in self.translate_dict:
```
Comment (mkurek) [inline metrology/reporter/statsd.py]: so now `translate_dict` is not used anywhere here?

File: metrology/reporter/statsd.py (hunk 2 above)
Comment (romcheg) [inline metrology/reporter/statsd.py]: Done

File: metrology/reporter/statsd.py (hunk 3 above)
Comment (romcheg) [inline metrology/reporter/statsd.py]: Will be resolved in the next change set

File: metrology/reporter/statsd.py (hunk 1 above)
Comment (romcheg) [inline metrology/reporter/statsd.py]: Done

**Comment by misiek08** [comment]
IMHO there are few unused imports. Then LGTM.

File: metrology/reporter/statsd.py [hunk 4]
```diff
@@ -51,40 +55,66 @@ def socket(self):
...
-            mtype = self.translate_dict[mtype]
-        else:
-            mtype = 'g'
 
-        self.batch_buffer += '{0}:{1}|{2}\n'.format(
-            name, value, mtype
-        )
+        m_type = self.translate_dict.get(mtype) or 'g'
+        metric =  '{0}:{1}|{2}\n'.format(name, value, mtype)
```
Comment (mkurek) [inline metrology/reporter/statsd.py]: so `mtype` or `m_type`?

File: metrology/reporter/statsd.py [hunk 5]
```diff
@@ -1,3 +1,7 @@
+from datetime import datetime
+import functools
+import os
+import re
```
Comment (mkurek) [inline metrology/reporter/statsd.py]: remove not used imports

File: metrology/reporter/statsd.py [hunk 6]
```diff
@@ -51,40 +55,66 @@ def socket(self):
...
-            self._buffered_send_metric(name, value, mtype)
+    def send_metric(self, name, mtype, metric, keys, snapshot_keys=None):
+        snapshot_keys = snapshot_keys or []
 
-    def _buffered_send_metric(self, name, value, mtype):
         if self.prefix:
-            name = '{0}.{1}'.format(self.prefix, name)
+            name = "{0}.{1}".format(self.prefix, name)
+
+        for name in keys:
```
Comment (mkurek) [inline metrology/reporter/statsd.py]: by using `name` in this for loop, you're overwriting `name` variable defined in line 98, so `self.prefix` and `name` of the metric is not used at all (only name of values, like count, rate etc)

File: metrology/reporter/statsd.py (hunk 4 above)
Comment (mkurek) [inline metrology/reporter/statsd.py]: btw use only one space after `=`

File: metrology/reporter/statsd.py [hunk 7]
```diff
@@ -51,40 +55,66 @@ def socket(self):
...
-        value = getattr(metric, keys, None)
-        if value:
-            self._buffered_send_metric(name, value, mtype)
+    def send_metric(self, name, mtype, metric, keys, snapshot_keys=None):
+        snapshot_keys = snapshot_keys or []
 
-    def _buffered_send_metric(self, name, value, mtype):
         if self.prefix:
-            name = '{0}.{1}'.format(self.prefix, name)
+            name = "{0}.{1}".format(self.prefix, name)
```
Comment (mkurek) [inline metrology/reporter/statsd.py]: what if name contains whitespaces? graphite reporter substitute it to underscore (https://github.com/cyberdelia/metrology/blob/master/metrology/reporter/graphite.py#L97)

Comment (mkurek) [inline metrology/reporter/statsd.py]: And, again, this method should be copied 1:1 from GraphiteReporter - so think about inheriting from GraphiteReporter and only overwriting `_buffered_send_metric` method (+ `_send_tcp`, `_send_udp` and `socket`)

**Comment by mkurek** [comment]
@cyberdelia what do you think about this? Is it ready to merge now?

**Comment by cyberdelia** [comment]
Thanks :bowtie:

**Pull request merged**